pandas/core/frame.py

```diff
@@ -5400,8 +5402,13 @@ def _homogenize(data, index, dtype=None):
                 v = v.reindex(index, copy=False)
         else:
             if isinstance(v, dict):
-                v = _dict_compat(v)
-                oindex = index.astype('O')
+                if oindex is None:
+                    oindex = index.astype('O')
+
+                if isinstance(index, (DatetimeIndex, TimedeltaIndex)):
```

Comment:
I don't remember why we do this, but is ``PeriodIndex`` needed as well? (and do we have tests for same)?